### DISCUSSION TITLE
Should FlatGeobuf support features with null geometries when spatially indexed?

### DISCUSSION THREAD

**Discussion by bjornharrtell** [flatgeobuf/flatgeobuf#260] [Q&A]
FlatGeobuf supports null geometries in the spec in some capacity but it has not been well explored.

Recently it was discovered that it doesn't currently seem possible to support the specific case when you want spatial indexed FlatGeobuf and the same time want it to contain features without geometry. See https://github.com/OSGeo/gdal/issues/7401 for more details.

In short, it is believed that since the header has a feature count that is also used to establish the index size and structure it must represent indexable features. Since features without geometry cannot be indexed there is a problem.

The FlatGeobuf spec can be extended with backwards compatibility if done carefully and I believe there could be a way to add the needed metainformation to support this case but I'm hesistant to do it / invest time into it since I don't personally like the mixed concern of non spatial and spatial features in the same file or feel I have use for it myself.

**Comment by alexbodn** [comment]
definitely, it's nice to keep all related data in one place, 
even not strictly geometric data.
without monstruous data duplication, a point, or a tl, dr 
multipoint could be autoindexed at the same location.
the user should be responsible of selecting the right records
for the right task.

**Reply by bjornharrtell** [reply]
I don't think I fully understand your comment.

**Reply by palmerj** [reply]
I also think FlatGeobuf files should support rows without a geometry, even if there is a spatial index. We have a few layers like this and still would like the option to publish the data as a single logical dataset, even if the spatial index request doesn't return that data. Note this also matches what is possible with most other spatial file or database stores.

**Reply by bjornharrtell** [reply]
I'm still hesistant, but perhaps more importantly still not interested so would probably need someone else to champion it.

As for other formats/databases supporting it I'm not sure is the best argument. For example, I find it rather annoying when null geometry has been left allowed in PostgreSQL/PostGIS more often than not by mistake and some crap data has sneaked its way in.

Disclaimer. I really dislike null in general. :D

**Reply by palmerj** [reply]
Up to you. But it's a well known design pattern that is accepted in the relational data world. Also the OGC simple features model allows it too, and OGR also implements that. So you will be going all of this and interop will decrease for your format.

**Reply by bjornharrtell** [reply]
That's a good argument that I agree with.

**Reply by alexbodn** [reply]
After reading the other comments, I feel I might have misunderstood your original post.
My vision was of one file containing geometries and let's say some comments, also localized.
Now I understand you meant one file containing spatially indexed geometries and non spatially indexed data.
It may lead to some fictional geometry designed not to interfere with spatial queries.
The problem is, this fictional data may be differ by application, such that the null value may still get a better, unified exception from index operations.

**Comment by bjornharrtell** [comment]
Revisited thought about this.. and perhaps all that is needed is an aux feature count for the features with null geometries.

**Reply by mtravis** [reply]
This would be a welcome addition. I've recently had issues when trying to convert parquet to FGB using DuckDB and the source data had tiny polygons that DuckDB reads as having no geometry.

**Reply by bjornharrtell** [reply]
Ew! Data destruction! 😱😂

**Comment by olivierdalang** [comment]
Just a little motivator for this one as it's the only drawback I know of flatgeobufs in read-only scenarios. If this was fixed we'd finally know what to tell to GIS users asking what format they should use :-D

**Reply by bjornharrtell** [reply]
It's too bad christmas is already over though. :)
